### DISCUSSION TITLE
Integration with/docs on react-three-fiber

### DISCUSSION THREAD

**Discussion by bergkvist** [uber/react-map-gl#1375] [General]
Right now, we have:

| imperative | declarative (react) |
|-|-|
| Mapbox GL JS | react-map-gl | 
| THREE.js | react-three-fiber |

Just as react-map-gl is the declarative react(ion) to Mapbox GL JS, react-three-fiber is the declarative reaction to THREE.js.

Mapbox GL JS has support for adding custom layers with THREE.js:
https://docs.mapbox.com/mapbox-gl-js/example/add-3d-model/

It would be really cool with some examples/documentation on how react-map-gl could be used together with react-three-fiber.

## Imagine being able to easily add a postprocessing effect to the map
```jsx
function App() {
  return (
    <Canvas>
      <ReactMapGL />
      <BookehPostprocessingEffect />
    </Canvas>
  )
}

```

## Or adding a custom 3d object
```jsx
function App() {
  return (
    <Canvas>
      <ReactMapGL />
      <Custom3DObject />
    </Canvas>
  )
}
```

**Comment by ibgreen** [comment]
react-map-gl is designed to work seamlessly with deck.gl (which by the way does support GPU based post processing effects).

That said, if someone contributed  good example on how to integrate with other 3D/WebGL libraries, I am sure it would be appreciated, but it would not be a priority for current maintainers.

**Comment by patrik-u** [comment]
Would love to see this as well.

**Comment by vincent-lecrubier-skydio** [comment]
I confirm, would love to see this, especially the ability to add custom 3d objects using react-three-fiber to mapbox gl stuff 👍

**Comment by hayk-skydio** [comment]
+1

**Comment by KevinGhadyani-Cityzenith** [comment]
This would greatly simplify needing to know a bunch of 3D stuff just to render pieces onto the map. Right now, you still need to know a lot about `three.js` to work with `mapbox-gl`.

**Comment by jo-chemla** [comment]
Same here ! Would allow us to embed massive 3d scans (up to ten billions of tris/pts primitives per scan) not only to a standard threejs scene via r3f (via loaders.gl [tiled loader](https://loaders.gl/docs/specifications/category-3d-tiles) for 3d-tiles or potree) or [nasa-3dtilesrenderer](https://github.com/NASA-AMMOS/3DTilesRendererJS) but also directly onto a map component for a seamless transition between the geospatial context and high-res 3d view

**Comment by RodrigoHamuy** [comment]
I built an NPM package called [react-three-map](https://github.com/RodrigoHamuy/react-three-map) to use `react-three-fiber` inside Mapbox/Maplibre the declarative way.

Even the README is inspired by this discussion thread :sweat_smile:  .

[![demo](https://github.com/RodrigoHamuy/react-three-map/raw/main/docs/basic-app.gif)](https://github.com/RodrigoHamuy/react-three-map)
